presto-raptorx/src/main/java/com/facebook/presto/raptorx/storage/ColumnStatsBuilder.java

```diff
@@ -0,0 +1,237 @@
+import static com.facebook.presto.spi.type.Varchars.isVarcharType;
+import static java.lang.Float.intBitsToFloat;
+import static java.lang.Math.toIntExact;
+
+public final class ColumnStatsBuilder
+{
+    /**
+     * Maximum length of a binary value stored in an index.
+     */
+    public static final int MAX_BINARY_INDEX_SIZE = 50;
+
+    private ColumnStatsBuilder() {}
+
+    public static byte[] truncateIndexValue(byte[] value)
+    {
+        if (value.length > MAX_BINARY_INDEX_SIZE) {
+            return Arrays.copyOfRange(value, 0, MAX_BINARY_INDEX_SIZE);
+        }
+        return value;
+    }
+
+    public static Slice truncateIndexValue(Slice slice)
+    {
+        if (slice.length() > MAX_BINARY_INDEX_SIZE) {
+            return slice.slice(0, MAX_BINARY_INDEX_SIZE);
+        }
+        return slice;
+    }
+
+    public static Optional<ColumnStats> computeColumnStats(OrcReader orcReader, long columnId, Type type)
+            throws IOException
+    {
+        return Optional.ofNullable(doComputeColumnStats(orcReader, columnId, type));
+    }
+
+    private static ColumnStats doComputeColumnStats(OrcReader orcReader, long columnId, Type type)
+            throws IOException
+    {
+        int columnIndex = columnIndex(orcReader.getColumnNames(), columnId);
+        OrcRecordReader reader = createOrcRecordReader(orcReader, ImmutableMap.of(columnIndex, type));
+
+        if (type.equals(BOOLEAN)) {
+            return indexBoolean(type, reader, columnIndex, columnId);
+        }
+        if (type.equals(BIGINT) ||
+                type.equals(INTEGER) ||
+                type.equals(SMALLINT) ||
+                type.equals(TINYINT) ||
+                type.equals(DATE) ||
+                type.equals(TIME) ||
+                type.equals(TIMESTAMP)) {
+            return indexLong(type, reader, columnIndex, columnId, Type::getLong);
+        }
+        if (type.equals(DOUBLE)) {
+            return indexLong(type, reader, columnIndex, columnId, ColumnStatsBuilder::readDouble);
+        }
+        if (type.equals(REAL)) {
+            return indexLong(type, reader, columnIndex, columnId, ColumnStatsBuilder::readReal);
+        }
+        if (isVarcharType(type) || type.equals(VARBINARY)) {
+            return indexBinary(type, reader, columnIndex, columnId);
+        }
+        return null;
+    }
+
+    private static int columnIndex(List<String> columnNames, long columnId)
+    {
+        int index = columnNames.indexOf(String.valueOf(columnId));
+        if (index == -1) {
+            throw new PrestoException(RAPTOR_INTERNAL_ERROR, "Missing column ID: " + columnId);
+        }
+        return index;
+    }
+
+    private static ColumnStats indexBoolean(Type type, OrcRecordReader reader, int columnIndex, long columnId)
+            throws IOException
+    {
+        boolean set = false;
+        boolean min = false;
+        boolean max = false;
+
+        while (true) {
```

Comment:
"highker on Feb 4  Member
This is going to read the whole file.... Why not using OrcMetadataReader?"

This is really bad. I wanted to change it 9 months ago....... Doesn't mean it's not bad. Just it's the same as  V1 and there for maybe 3 years. 

I would put it in the above mentioned "refactor whole writing path" Task, including removing things like orcFileWriter, and change here "not read whole file back".

Though I don't expect performance improving. I remember Linux flush dirty page every 5 seconds and only flush those older than 30 seconds, and only after that, "read back" has possibility to hit disk (just possibility, not certain).